docs/load_balancer.md

```diff
@@ -0,0 +1,278 @@
+
+The default value is `--public_service_ports=10000-20000`.  This excludes most well-known ports, and certainly excludes
+the <1024 ports.  We should try to keep the minion using ports <1024.
+
+TODO: figure out which >1024 ports are needed on the minion
+
+A null range can be specified (e.g. `--public_service_ports=0-0`).  This will have the effect of preventing any public port allocations.
+
+In future, we could allow disjoint port ranges (`--public_service_ports=10000-20000,22000-30000`), but not for V1
+
+
+## API evolution
+
+The main changes:
+
+* we deprecate createExternalLoadBalancer, and replace it with visibility
+* we move publicIPs to the Status
+* we add a field hostPort to each service port
+ 
+Mappings:
+
+`visibility` is mappable:
+
+```
+new.visibility = old.createExternalLoadBalancer ? "loadbalancer" : "cluster"
+```
+
+```
+old.createExternalLoadBalancer = (new.visibility == 'loadbalancer')
+```
+
+`publicIPs` will be copied to the Spec if set on the Status, and from the Spec to the Status if not set on the Status.
+Once each cloud provider sets the Status, we will stop copying from the Spec to the Status.
+
+For v1 final:
+* we should remove publicIPs from the Spec, and require that this information is retrieved through the Status.
+However, we would need to change the openstack code to auto-allocate a VIP.
+* we should remove createExternalLoadBalancer.
+
+## kube-proxy pseudocode
+
+```
+for each service {
+    for each service.port {
+        // listen on cluster
+        listen on <internalMagicIp>:0
+        proxyPort = figureOutTheRandomPortWeJustGotAssigned()
+        openPortal(service.portalIP, port.port, proxyPort)
+        
+
+        if service.spec.visibility == "loadbalancer" || (service.spec.visibility == "" && service.createExternalLoadBalancer=true) {
+          if (port.hostPort != 0) {
+            openLoadBalancerPortals(service, port.port, port.hostPort) // uses cloud-provider to figure out correct iptables
+          }
+        } else if service.spec.visibility == "public" {
+          if (port.hostPort != 0) {
+            openPublicPortal(<publicMagicIp>, port.port, port.hostPort)
+          }
+        } else if service.spec.visibility == "" {
+          for each service.publicIP {  // deprecated
+            openPublicPortal(publicIP, port.port, proxyPort) // uses cloud-provider to decide srcip vs dstip (gclb vs rax)
+          }
+        }
+   }
+}
+```
+
+
+## kube-proxy iptables
+
+As part of the implementation of kube-proxy's internal magic, kube-proxy allocates ports that are used locally by
+iptables.  These ports are not visible other than to kube-proxy & iptables.  We need to be sure that these ports will
+not conflict with the cluster-wide ports.  To do that, we must make sure that kube-proxy differentiates between
+internal & external traffic, and uses these internal ports only with internal traffic.
+
+There are a few possible implementation strategies here:
```

Comment:
This whole section needs a revamp, yeah?  I think we agreed to leave the port allocation pretty much as-is.  random ports for internal services and stable ports for public services.